# Aviation Accidents Analysis

You are part of a consulting firm that is tasked to do an analysis of commercial and passenger jet airline safety. The client (an airline/airplane insurer) is interested in knowing what types of aircraft (makes/models) exhibit low rates of total destruction and low likelihood of fatal or serious passenger injuries in the event of an accident. They are also interested in any general variables/conditions that might be at play. Your analysis will be based off of aviation accident data accumulated from the years 1948-2023. 

Our client is only interested in airplane makes/models that are professional builds and could potentially still be active. Assume a max lifetime of 40 years for a make/model retirement and make sure to filter your data accordingly (i.e. from 1983 onwards). They would also like separate recommendations for small aircraft vs. larger passenger models. **In addition, make sure that claims that you make are statistically robust and that you have enough samples when making comparisons between groups.**


In this summative assessment you will demonstrate your ability to:
- **Use Pandas to load, inspect, and clean the dataset appropriately.**
- **Transform relevant columns to create measures that address the problem at hand.**
- conduct EDA: visualization and statistical measures to systematically understand the structure of the data
- recommend a set of airplanes and makes conforming to the client's request and identify at least *two* factors contributing to airplane safety. You must provide supporting evidence (visuals, summary statistics, tables) for each claim you make.

### Make relevant library imports

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings('ignore')

## Data Loading and Inspection

### Load in data from the relevant directory and inspect the dataframe.
- inspect NaNs, datatypes, and summary statistics

In [2]:
# Load the csv 
df = pd.read_csv("AviationData.csv", index_col=0, encoding='latin1')
df.shape

(88889, 30)

Observation:The dataset has 88,889 rows and 30 columns.

In [3]:
# Check top 5 rows
df.head()

,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,Injury.Severity,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
Event.Id,,,,,,,,,,,,,,,,,,,,,
20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,Fatal(2),...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,Fatal(4),...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,Fatal(3),...,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,Fatal(2),...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,Fatal(1),...,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980


In [4]:
# Check bottom 5 rows
df.tail()

,Investigation.Type,Accident.Number,Event.Date,Location,Country,Latitude,Longitude,Airport.Code,Airport.Name,Injury.Severity,...,Purpose.of.flight,Air.carrier,Total.Fatal.Injuries,Total.Serious.Injuries,Total.Minor.Injuries,Total.Uninjured,Weather.Condition,Broad.phase.of.flight,Report.Status,Publication.Date
Event.Id,,,,,,,,,,,,,,,,,,,,,
20221227106491,Accident,ERA23LA093,2022-12-26,"Annapolis, MD",United States,NaN,NaN,NaN,NaN,Minor,...,Personal,NaN,0.0,1.0,0.0,0.0,NaN,NaN,NaN,29-12-2022
20221227106494,Accident,ERA23LA095,2022-12-26,"Hampton, NH",United States,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
20221227106497,Accident,WPR23LA075,2022-12-26,"Payson, AZ",United States,341525N,1112021W,PAN,PAYSON,Non-Fatal,...,Personal,NaN,0.0,0.0,0.0,1.0,VMC,NaN,NaN,27-12-2022
20221227106498,Accident,WPR23LA076,2022-12-26,"Morgan, UT",United States,NaN,NaN,NaN,NaN,NaN,...,Personal,MC CESSNA 210N LLC,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
20221230106513,Accident,ERA23LA097,2022-12-29,"Athens, GA",United States,NaN,NaN,NaN,NaN,Minor,...,Personal,NaN,0.0,1.0,0.0,1.0,NaN,NaN,NaN,30-12-2022


In [5]:
# Check datatypes and info
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 88889 entries, 20001218X45444 to 20221230106513
Data columns (total 30 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Investigation.Type      88889 non-null  object 
 1   Accident.Number         88889 non-null  object 
 2   Event.Date              88889 non-null  object 
 3   Location                88837 non-null  object 
 4   Country                 88663 non-null  object 
 5   Latitude                34382 non-null  object 
 6   Longitude               34373 non-null  object 
 7   Airport.Code            50132 non-null  object 
 8   Airport.Name            52704 non-null  object 
 9   Injury.Severity         87889 non-null  object 
 10  Aircraft.damage         85695 non-null  object 
 11  Aircraft.Category       32287 non-null  object 
 12  Registration.Number     87507 non-null  object 
 13  Make                    88826 non-null  object 
 14  Model                

In [6]:
# Check columns
df.columns

Index(['Investigation.Type', 'Accident.Number', 'Event.Date', 'Location',
       'Country', 'Latitude', 'Longitude', 'Airport.Code', 'Airport.Name',
       'Injury.Severity', 'Aircraft.damage', 'Aircraft.Category',
       'Registration.Number', 'Make', 'Model', 'Amateur.Built',
       'Number.of.Engines', 'Engine.Type', 'FAR.Description', 'Schedule',
       'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

In [7]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Number.of.Engines,82805.0,1.146585,0.446510,0.0,1.0,1.0,1.0,8.0
Total.Fatal.Injuries,77488.0,0.647855,5.485960,0.0,0.0,0.0,0.0,349.0
Total.Serious.Injuries,76379.0,0.279881,1.544084,0.0,0.0,0.0,0.0,161.0
Total.Minor.Injuries,76956.0,0.357061,2.235625,0.0,0.0,0.0,0.0,380.0
Total.Uninjured,82977.0,5.325440,27.913634,0.0,0.0,1.0,2.0,699.0


In [8]:
# Check missing values
missing_percentages = (df.isnull().sum() / len(df)) * 100
missing_percentages.sort_values(ascending=False)

Schedule                  85.845268
Air.carrier               81.271023
FAR.Description           63.974170
Aircraft.Category         63.677170
Longitude                 61.330423
Latitude                  61.320298
Airport.Code              43.601570
Airport.Name              40.708074
Broad.phase.of.flight     30.560587
Publication.Date          15.492356
Total.Serious.Injuries    14.073732
Total.Minor.Injuries      13.424608
Total.Fatal.Injuries      12.826109
Engine.Type                7.982990
Report.Status              7.181991
Purpose.of.flight          6.965991
Number.of.Engines          6.844491
Total.Uninjured            6.650992
Weather.Condition          5.053494
Aircraft.damage            3.593246
Registration.Number        1.554748
Injury.Severity            1.124999
Country                    0.254250
Amateur.Built              0.114750
Model                      0.103500
Make                       0.070875
Location                   0.058500
Accident.Number            0

Observation:Columns like schedule, air.carrier, and far.description have very high missing percentages above 50% and may need to be dropped.

## Data Cleaning

### Filtering aircrafts and events

We want to filter the dataset to include aircraft that the client is interested in an analysis of:
- inspect relevant columns
- figure out any reasonable imputations
- filter the dataset

In [9]:
# Make a copy
airline_df = df.copy(deep=True)

In [10]:
# Remove white space
airline_df.columns = df.columns.str.replace(".", "_", regex=False).str.lower().str.strip()
airline_df

,investigation_type,accident_number,event_date,location,country,latitude,longitude,airport_code,airport_name,injury_severity,...,purpose_of_flight,air_carrier,total_fatal_injuries,total_serious_injuries,total_minor_injuries,total_uninjured,weather_condition,broad_phase_of_flight,report_status,publication_date
Event.Id,,,,,,,,,,,,,,,,,,,,,
20001218X45444,Accident,SEA87LA080,1948-10-24,"MOOSE CREEK, ID",United States,NaN,NaN,NaN,NaN,Fatal(2),...,Personal,NaN,2.0,0.0,0.0,0.0,UNK,Cruise,Probable Cause,NaN
20001218X45447,Accident,LAX94LA336,1962-07-19,"BRIDGEPORT, CA",United States,NaN,NaN,NaN,NaN,Fatal(4),...,Personal,NaN,4.0,0.0,0.0,0.0,UNK,Unknown,Probable Cause,19-09-1996
20061025X01555,Accident,NYC07LA005,1974-08-30,"Saltville, VA",United States,36.922223,-81.878056,NaN,NaN,Fatal(3),...,Personal,NaN,3.0,NaN,NaN,NaN,IMC,Cruise,Probable Cause,26-02-2007
20001218X45448,Accident,LAX96LA321,1977-06-19,"EUREKA, CA",United States,NaN,NaN,NaN,NaN,Fatal(2),...,Personal,NaN,2.0,0.0,0.0,0.0,IMC,Cruise,Probable Cause,12-09-2000
20041105X01764,Accident,CHI79FA064,1979-08-02,"Canton, OH",United States,NaN,NaN,NaN,NaN,Fatal(1),...,Personal,NaN,1.0,2.0,NaN,0.0,VMC,Approach,Probable Cause,16-04-1980
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20221227106491,Accident,ERA23LA093,2022-12-26,"Annapolis, MD",United States,NaN,NaN,NaN,NaN,Minor,...,Personal,NaN,0.0,1.0,0.0,0.0,NaN,NaN,NaN,29-12-2022
20221227106494,Accident,ERA23LA095,2022-12-26,"Hampton, NH",United States,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,0.0,0.0,0.0,0.0,NaN,NaN,NaN,NaN
20221227106497,Accident,WPR23LA075,2022-12-26,"Payson, AZ",United States,341525N,1112021W,PAN,PAYSON,Non-Fatal,...,Personal,NaN,0.0,0.0,0.0,1.0,VMC,NaN,NaN,27-12-2022


Observation:Column names have been changed to lowercase with underscores replacing dots, making them easier to work with.

In [11]:
# Check unique values in aircraft_category
airline_df['aircraft_category'].value_counts()

aircraft_category
Airplane             27617
Helicopter            3440
Glider                 508
Balloon                231
Gyrocraft              173
Weight-Shift           161
Powered Parachute       91
Ultralight              30
Unknown                 14
WSFT                     9
Powered-Lift             5
Blimp                    4
UNK                      2
Rocket                   1
ULTR                     1
Name: count, dtype: int64

In [12]:
# Filter Airplanes 
airline_df = airline_df[airline_df['aircraft_category'] == 'Airplane']

# Check
airline_df.shape

(27617, 30)

In [13]:
# Check unique values in amateur_built
airline_df['amateur_built'].value_counts()

amateur_built
No     24417
Yes     3183
Name: count, dtype: int64

In [14]:
# Check event_date range
airline_df['event_date'].value_counts()

event_date
1982-05-16    23
1982-07-09    21
1982-06-19    20
1982-04-18    20
1982-03-13    19
              ..
2022-10-30     1
2022-12-02     1
2022-12-01     1
2022-11-29     1
1982-01-16     1
Name: count, Length: 7328, dtype: int64

In [15]:
# Filter for safe flights
airline_df = airline_df[airline_df['amateur_built'] == 'No']
airline_df.shape

(24417, 30)

In [16]:
# Convert event_date to datetime
airline_df['event_date'] = pd.to_datetime(airline_df['event_date'])

In [17]:
# Filter flights from 1983
airline_df = airline_df[airline_df['event_date'].dt.year >= 1983]
airline_df.shape

(21447, 30)

### Cleaning and constructing Key Measurables

Injuries and robustness to destruction are a key interest point for the client. Clean and impute relevant columns and then create derived fields that best quantifies what the client wishes to track. **Use commenting or markdown to explain any cleaning assumptions as well as any derived columns you create.**

**Construct metric for fatal/serious injuries**

*Hint:* Estimate the total number of passengers on each flight. The likelihood of serious / fatal injury can be estimated as a fraction from this.

In [18]:
# Check injury columns
airline_df[['total_fatal_injuries', 'total_serious_injuries', 'total_minor_injuries', 'total_uninjured']].head(10)

,total_fatal_injuries,total_serious_injuries,total_minor_injuries,total_uninjured
Event.Id,,,,
20001214X42478,NaN,NaN,NaN,588.0
20001214X42478,NaN,NaN,NaN,588.0
20001214X42331,1.0,1.0,NaN,NaN
20001214X42672,1.0,NaN,NaN,4.0
20001214X44248,NaN,NaN,NaN,289.0
20001214X44100,11.0,2.0,NaN,13.0
20001214X44944,1.0,NaN,NaN,29.0
20001214X45013,NaN,NaN,NaN,100.0
20001214X45188,NaN,NaN,NaN,1.0


In [19]:
# Fill NaN values with 0 for injury columns
airline_df['total_fatal_injuries'] = airline_df['total_fatal_injuries'].fillna(0)
airline_df['total_serious_injuries'] = airline_df['total_serious_injuries'].fillna(0)
airline_df['total_minor_injuries'] = airline_df['total_minor_injuries'].fillna(0)
airline_df['total_uninjured'] = airline_df['total_uninjured'].fillna(0)
airline_df

,investigation_type,accident_number,event_date,location,country,latitude,longitude,airport_code,airport_name,injury_severity,...,purpose_of_flight,air_carrier,total_fatal_injuries,total_serious_injuries,total_minor_injuries,total_uninjured,weather_condition,broad_phase_of_flight,report_status,publication_date
Event.Id,,,,,,,,,,,,,,,,,,,,,
20001214X42478,Incident,LAX83IA149B,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,LAX,LOS ANGELES INTL,Incident,...,Unknown,NaN,0.0,0.0,0.0,588.0,VMC,Standing,Probable Cause,04-12-2014
20001214X42478,Incident,LAX83IA149A,1983-03-18,"LOS ANGELES, CA",United States,NaN,NaN,LAX,LOS ANGELES INTL,Incident,...,NaN,"Singapore Airlines, Ltd.",0.0,0.0,0.0,588.0,VMC,Taxi,Probable Cause,04-12-2014
20001214X42331,Accident,ATL83FA140,1983-03-20,"CROSSVILLE, TN",United States,NaN,NaN,NaN,NaN,Fatal(1),...,Personal,NaN,1.0,1.0,0.0,0.0,IMC,Cruise,Probable Cause,02-05-2011
20001214X42672,Accident,FTW83LA177,1983-04-02,"MCKINNEY, TX",United States,NaN,NaN,TX05,AERO COUNTRY,Fatal(1),...,Skydiving,NaN,1.0,0.0,0.0,4.0,VMC,Standing,Probable Cause,17-10-2016
20001214X44248,Incident,MIA83IA210,1983-08-21,"NORFOLK, VA",United States,NaN,NaN,NaN,NaN,Incident,...,Unknown,NaN,0.0,0.0,0.0,289.0,VMC,Cruise,Probable Cause,01-02-2016
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
20221213106455,Accident,WPR23LA065,2022-12-13,"Lewistown, MT",United States,047257N,0109280W,KLWT,Lewiston Municipal Airport,Non-Fatal,...,NaN,NaN,0.0,0.0,0.0,1.0,NaN,NaN,NaN,14-12-2022
20221215106463,Accident,ERA23LA090,2022-12-14,"San Juan, PR",United States,182724N,0066554W,SIG,FERNANDO LUIS RIBAS DOMINICCI,Non-Fatal,...,Personal,SKY WEST AVIATION INC TRUSTEE,0.0,0.0,0.0,1.0,VMC,NaN,NaN,27-12-2022
20221219106475,Accident,WPR23LA069,2022-12-15,"Wichita, KS",United States,373829N,0972635W,ICT,WICHITA DWIGHT D EISENHOWER NT,Non-Fatal,...,NaN,NaN,0.0,0.0,0.0,1.0,NaN,NaN,NaN,19-12-2022


Observation:Missing values in injury columns have been filled with 0, assuming no injuries were recorded.

In [20]:
# number of passengers on each flight
airline_df['total_passengers'] = (airline_df['total_fatal_injuries'] + 
                                   airline_df['total_serious_injuries'] + 
                                   airline_df['total_minor_injuries'] + 
                                   airline_df['total_uninjured'])
airline_df['total_passengers'].head()

Event.Id
20001214X42478    588.0
20001214X42478    588.0
20001214X42331      2.0
20001214X42672      5.0
20001214X44248    289.0
Name: total_passengers, dtype: float64

In [21]:
# Calculate fraction of passengers seriously/fatally injured
airline_df['injury_fraction'] = (airline_df['total_fatal_injuries'] + 
                                  airline_df['total_serious_injuries']) / airline_df['total_passengers']

In [22]:
# Check the new column
airline_df['injury_fraction'].describe()

count    20543.000000
mean         0.283971
std          0.431704
min          0.000000
25%          0.000000
50%          0.000000
75%          0.800000
max          1.000000
Name: injury_fraction, dtype: float64

Observation:The mean injury fraction is 0.28, meaning on average 28% of passengers are seriously or fatally injured in an accident.

**Aircraft.Damage**
- identify and execute any cleaning tasks
- construct a derived column tracking whether an aircraft was destroyed or not.

In [23]:
# Check unique values in aircraft_damage
airline_df['aircraft_damage'].value_counts()

aircraft_damage
Substantial    16990
Destroyed       2316
Minor            817
Unknown           97
Name: count, dtype: int64

In [24]:
# Replace Unknown with NaN
airline_df['aircraft_damage'] = airline_df['aircraft_damage'].replace('Unknown', np.nan)
airline_df['aircraft_damage'].value_counts()

aircraft_damage
Substantial    16990
Destroyed       2316
Minor            817
Name: count, dtype: int64

Observation:Unknown values in aircraft_damage have been replaced with NaN.

In [25]:
# Create is_destroyed column
airline_df['is_destroyed'] = airline_df['aircraft_damage'].map({'Destroyed': 1, 'Substantial': 0, 'Minor': 0})
airline_df['is_destroyed'].value_counts()

is_destroyed
0.0    17807
1.0     2316
Name: count, dtype: int64

Observation:2,316 aircraft were completely destroyed.

### Investigate the *Make* column
- Identify cleaning tasks here
- List cleaning tasks clearly in markdown
- Execute the cleaning tasks
- For your analysis, keep Makes with a reasonable number (you can put the threshold at 50 though lower could work as well)

In [26]:
# Check unique values in make column
airline_df['make'].value_counts()

make
CESSNA               4867
PIPER                2803
Cessna               2279
Piper                1186
BOEING               1037
                     ... 
Believer                1
KAKELDEY RONALD O       1
CASSUTT                 1
Cassult Racer           1
GEIST DONALD R          1
Name: count, Length: 1332, dtype: int64

In [27]:
# Capitalize the first letter of each
airline_df['make'] = airline_df['make'].str.strip().str.title()

# Check it
airline_df['make'].value_counts()

make
Cessna                            7146
Piper                             3989
Beech                             1431
Boeing                            1264
Mooney                             363
                                  ... 
Aeroprakt Manufacturing Sp Zoo       1
Mcglaughlin Richard                  1
Aeros                                1
Orlican S R O                        1
Kuykendall Charles                   1
Name: count, Length: 1087, dtype: int64

Observation:duplicate entries like CESSNA and Cessna have been merged.

In [28]:
# Keep only makes with 50 or more occurrences
make_counts = airline_df['make'].value_counts()
valid_makes = make_counts[make_counts >= 50].index

airline_df = airline_df[airline_df['make'].isin(valid_makes)]

# Check shape
airline_df.shape

(17892, 33)

Observation:After keeping only makes with 50+ records, dataset reduced to 17,892 rows.

In [29]:
# Check model column
airline_df['model'].value_counts()

model
172            769
737            403
152            316
182            304
172S           276
              ... 
305C (0-1E)      1
7GCBA            1
M-4-210          1
AT-802F          1
737-9            1
Name: count, Length: 2034, dtype: int64

### Inspect Model column
- Get rid of any NaNs.
- Inspect the column and counts for each model/make. Are model labels unique to each make?
- If not, create a derived column that is a unique identifier for a given plane type.

In [30]:
# Check NaNs in model column
airline_df['model'].isna().sum()

np.int64(13)

In [31]:
# Drop NaN models
airline_df = airline_df.dropna(subset=['model'])

# Check
airline_df.shape

(17879, 33)

In [32]:
# Create unique identifier combining make and model
airline_df['make_model'] = airline_df['make'] + ' ' + airline_df['model']

# Check it
airline_df['make_model'].value_counts().head(10)

make_model
Cessna 172     769
Boeing 737     403
Cessna 152     316
Cessna 182     304
Cessna 172S    276
Piper PA28     273
Cessna 172N    249
Cessna 180     213
Cessna 172M    180
Cessna 150     179
Name: count, dtype: int64

Observation: Cessna 172 is the most common aircraft type involved in accidents with 769 records.

### Cleaning other columns
- there are other columns containing data that might be related to the outcome of an accident. We list a few here:
- Engine.Type
- Weather.Condition
- Number.of.Engines
- Purpose.of.flight
- Broad.phase.of.flight

Inspect and identify potential cleaning tasks in each of the above columns. Execute those cleaning tasks. 

**Note**: You do not necessarily need to impute or drop NaNs here.

In [33]:
# Check engine_type
airline_df['engine_type'].value_counts()

engine_type
Reciprocating      12835
Turbo Prop           931
Turbo Fan            701
Unknown              105
Turbo Jet             71
Geared Turbofan       12
Turbo Shaft            9
UNK                    1
Name: count, dtype: int64

In [34]:
# Replace UNK with Unknown
airline_df['engine_type'] = airline_df['engine_type'].replace('UNK', 'Unknown')

# Check it
airline_df['engine_type'].value_counts()

engine_type
Reciprocating      12835
Turbo Prop           931
Turbo Fan            701
Unknown              106
Turbo Jet             71
Geared Turbofan       12
Turbo Shaft            9
Name: count, dtype: int64

In [35]:
# Check weather_condition
airline_df['weather_condition'].value_counts()

weather_condition
VMC    14295
IMC      905
Unk      186
UNK       76
Name: count, dtype: int64

In [36]:
# Replace Unk and UNK with Unknown
airline_df['weather_condition'] = airline_df['weather_condition'].replace({'Unk': 'Unknown', 'UNK': 'Unknown'})

# Check it
airline_df['weather_condition'].value_counts()

weather_condition
VMC        14295
IMC          905
Unknown      262
Name: count, dtype: int64

Observation:UNK values changed to Unknown in engine_type and weather_condition columns.

In [37]:
# Check number_of_engines
airline_df['number_of_engines'].value_counts()

number_of_engines
1.0    13222
2.0     2470
4.0       67
3.0       26
0.0        5
Name: count, dtype: int64

In [38]:
# Replace 0 engines with NaN
airline_df['number_of_engines'] = airline_df['number_of_engines'].replace(0, np.nan)

# Check it
airline_df['number_of_engines'].value_counts()

number_of_engines
1.0    13222
2.0     2470
4.0       67
3.0       26
Name: count, dtype: int64

Observation:Invalid 0 engine entries replaced with NaN.

In [39]:
# Check purpose_of_flight
airline_df['purpose_of_flight'].value_counts()

purpose_of_flight
Personal                     9844
Instructional                2410
Aerial Application            724
Business                      409
Unknown                       303
Positioning                   269
Skydiving                     157
Aerial Observation            147
Other Work Use                121
Banner Tow                     86
Flight Test                    73
Ferry                          72
Executive/corporate            65
Glider Tow                     29
Public Aircraft - Federal      28
Public Aircraft                27
Public Aircraft - State        21
Air Race show                  15
Firefighting                   12
Public Aircraft - Local        11
PUBS                            3
Air Race/show                   2
Air Drop                        2
ASHO                            2
Name: count, dtype: int64

In [40]:
# Fix Air Race show
airline_df['purpose_of_flight'] = airline_df['purpose_of_flight'].replace('Air Race/show', 'Air Race show')

# Check
airline_df['purpose_of_flight'].value_counts()

purpose_of_flight
Personal                     9844
Instructional                2410
Aerial Application            724
Business                      409
Unknown                       303
Positioning                   269
Skydiving                     157
Aerial Observation            147
Other Work Use                121
Banner Tow                     86
Flight Test                    73
Ferry                          72
Executive/corporate            65
Glider Tow                     29
Public Aircraft - Federal      28
Public Aircraft                27
Public Aircraft - State        21
Air Race show                  17
Firefighting                   12
Public Aircraft - Local        11
PUBS                            3
Air Drop                        2
ASHO                            2
Name: count, dtype: int64

Observation:Air Race show and Air Race/show are the same thing so i changed it to Air Race Show.

In [41]:
# Check broad_phase_of_flight
airline_df['broad_phase_of_flight'].value_counts()

broad_phase_of_flight
Landing        1110
Takeoff         425
Cruise          238
Approach        210
Maneuvering     127
Taxi             99
Go-around        81
Descent          62
Climb            52
Standing         35
Unknown          11
Other             2
Name: count, dtype: int64

### Column Removal
- inspect the dataframe and drop any columns that have too many NaNs

In [42]:
# Check missing values again
missing_percentages = (airline_df.isnull().sum() / len(airline_df)) * 100
missing_percentages.sort_values(ascending=False)

schedule                  88.036244
broad_phase_of_flight     86.285586
air_carrier               52.749035
airport_code              34.850942
airport_name              34.258068
report_status             21.170088
engine_type               17.976397
purpose_of_flight         17.042340
weather_condition         13.518653
number_of_engines         11.712064
longitude                 10.632586
latitude                  10.615806
aircraft_damage            6.409754
is_destroyed               6.409754
publication_date           4.407405
injury_fraction            4.340287
injury_severity            4.010291
far_description            1.924045
registration_number        0.917277
location                   0.022373
country                    0.005593
investigation_type         0.000000
accident_number            0.000000
event_date                 0.000000
aircraft_category          0.000000
make                       0.000000
model                      0.000000
amateur_built              0

In [43]:
# Drop columns with more than 50% missing values
airline_df = airline_df.drop(columns=['schedule', 'air_carrier'])

# Check shape
airline_df.shape

(17879, 32)

Observation:Dropped schedule and air_carrier columns due to more than 50% missing values. Dataset now has 32 columns.

### Save DataFrame to csv
- its generally useful to save data to file/server after its in a sufficiently cleaned or intermediate state
- the data can then be loaded directly in another notebook for further analysis
- this helps keep your notebooks and workflow readable, clean and modularized

In [44]:
# Save cleaned data to csv
airline_df.to_csv('cleaned_aviation_data.csv')

print("Saved!")

Saved!
